### PR TITLE
Wip rgw sync fixes

### PR DESCRIPTION
**Pull request by yehudasa**

### CODE REVIEW

File: src/rgw/rgw_data_sync.cc [hunk 1]
```diff
@@ -1230,10 +1244,13 @@ class RGWDataSyncCR : public RGWCoroutine {
...
               RGWDataSyncShardControlCR *cr = new RGWDataSyncShardControlCR(sync_env, sync_env->store->get_zone_params().log_pool,
                                                         iter->first, iter->second);
+              cr->get();
               shard_crs_lock.Lock();
               shard_crs[iter->first] = cr;
               shard_crs_lock.Unlock();
-              spawn(cr, true);
+              RGWCoroutinesStack *stack = spawn(cr, true);
+              stack->get();
+              stacks[stack] = iter->first;
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc]: I'm not sure what `map<RGWCoroutinesStack *, int> stacks` accomplishes here (I don't see it being accessed elsewhere), and this reference from `stack->get()` isn't ever released.

File: src/rgw/rgw_coroutine.h [hunk 2]
```diff
@@ -271,6 +271,9 @@ class RGWCoroutine : public RefCountedObject, public boost::asio::coroutine {
   int wait(const utime_t& interval);
   bool drain_children(int num_cr_left); /* returns true if needed to be called again */
   void wakeup();
+  void restart() {
+    set_state(RGWCoroutine_Run);
+  }
```
Comment (cbodley) [inline src/rgw/rgw_coroutine.h]: I don't see this called anywhere

File: src/rgw/rgw_data_sync.cc (hunk 1 above)
Comment (yehudasa) [inline src/rgw/rgw_data_sync.cc]: Yeah, my original change was a bit more complicated, was tracking spawned crs, and then re-executing. But ended up with the current solution. Need to remove this line.

File: src/rgw/rgw_coroutine.h (hunk 2 above)
Comment (yehudasa) [inline src/rgw/rgw_coroutine.h]: yuo

**Comment by yehudasa** [comment]
@cbodley pushed a fix, merging it

**Comment by yehudasa** [comment]
@cbodley *fixed according to your comments (removed unneeded stuff)

**Pull request merged**